README.pod
=head1 NAME

Metabase::Backend::AWS - Metabase backend implemented using Amazon Web Services

=head1 VERSION

version 1.000

=head1 SYNOPSIS

XXX consolidate synopses from modules

=head1 DESCRIPTION

This distribution provides a backend for L<Metabase> using Amazon Web Services.
There are two modules included, L<Metabase::Index::SimpleDB> and
L<Metabase::Archive::S3>.  They can be used separately or together (see
L<Metabase::Librarian> for details).

The L<Metabase::Backend::AWX> module is a L<Moose::Role> that provides
common attributes and private helpers and is not intended to be used directly.

Common attributes are described further below.

=head1 ATTRIBUTES

=head2 access_key_id

An AWS Access Key ID

=head2 secret_access_key

An AWS Secret Access Key matching the Access Key ID

=for Pod::Coverage method_names_here

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<http://rt.cpan.org/Public/Dist/Display.html?Name=Metabase-Backend-AWS>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/metabase-backend-aws>

  git clone https://github.com/dagolden/metabase-backend-aws.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2011 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004



Changes
Revision history for Metabase-Backend-AWS

{{$NEXT}}

1.000     2012-05-16 13:41:41 America/New_York

    - Split out from the Metabase distribution



t/lib/Metabase/Test/Index/SimpleDB.pm
use 5.010;
use strict;
use warnings;
package Metabase::Test::Index::SimpleDB;

use Test::More;
use Test::Routine;
use Net::Amazon::Config;
use Metabase::Index::SimpleDB;

my $profile_env = "PERL_METABASE_TEST_AWS_PROFILE";
unless ( $ENV{$profile_env} ) {
  plan skip_all => "No \$ENV{$profile_env} provided for testing";
}

has amazon_config => (
  is => 'ro',
  isa => 'Net::Amazon::Config',
  default => sub { Net::Amazon::Config->new },
);

has 'profile' => (
  is      => 'ro',
  isa     => 'Net::Amazon::Config::Profile',
  lazy_build    => 1,
  handles => [ qw/access_key_id secret_access_key/ ],
);
 
has 'test_domain' => (
  is => 'ro',
  isa => 'Str',
  lazy_build => 1,
);

sub _build_test_domain {
  return "org.cpantesters.metabase.test" . int(rand(2**31));
}

sub _build_profile {
  my $self = shift;
  die "No \$ENV{$profile_env}\n" unless $ENV{$profile_env};
  return $self->amazon_config->get_profile( $ENV{$profile_env} );
}

sub _build_index {
  my $self = shift;
  return Metabase::Index::SimpleDB->new(
      access_key_id     => $self->access_key_id,
      secret_access_key => $self->secret_access_key,
      domain            => $self->test_domain,
      consistent        => 1,
  );
}

before clear_index => sub {
  my $self = shift;
  $self->index->simpledb->send_request(
    'DeleteDomain', { DomainName => $self->test_domain }
  );
};

sub DEMOLISH { my $self = shift; $self->clear_index; }

1;



lib/Metabase/Archive/S3.pm
use 5.006;
use strict;
use warnings;

package Metabase::Archive::S3;
# VERSION

use Moose;
use namespace::autoclean;
use Moose::Util::TypeConstraints;
use MooseX::Types::Path::Class;
use namespace::autoclean;

use Metabase::Fact;
use Carp       ();
use Data::GUID ();
use Data::Stream::Bulk::Filter 0.08;
use JSON 2 ();
use Net::Amazon::S3;
use Path::Class ();
use Compress::Zlib 2 qw(compress uncompress);

with 'Metabase::Backend::AWS';
with 'Metabase::Archive' => { -version => 1.000 };

# Prefix string must have a trailing slash but not leading slash
subtype 'PrefixStr'
  => as 'Str'
  => where { $_ =~ m{^\w} && $_ =~ m{/$} };

coerce 'PrefixStr'
  => from 'Str' => via { s{/$}{}; s{^/}{}; $_ . "/" };

has 'bucket' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has 'prefix' => (
    is       => 'ro',
    isa      => 'PrefixStr',
    required => 1,
    coerce   => 1,
);

has 'compressed' => (
    is      => 'rw',
    isa     => 'Bool',
    default => 1,
);

has 's3_bucket' => (
    is       => 'ro',
    isa      => 'Net::Amazon::S3::Client::Bucket',
    required => 1,
    lazy     => 1,
    default  => sub {
        my $self = shift;
        my $s3   = Net::Amazon::S3->new(
            aws_access_key_id     => $self->access_key_id,
            aws_secret_access_key => $self->secret_access_key,
            retry                 => 1,
        );
        my $client = Net::Amazon::S3::Client->new( s3 => $s3 );
        my $method = (grep { $_ eq $self->bucket } $client->buckets) ? 'bucket' : 'create_bucket';
        return $client->$method( name => $self->bucket );
    }
);

has '_json' => (
  is => 'ro',
  required => 1,
  lazy => 1,
  default => sub { JSON->new->ascii },
);

sub initialize {}

# given fact, store it and return guid;
sub store {
    my ( $self, $fact_struct ) = @_;
    my $guid = $fact_struct->{metadata}{core}{guid};
    my $type = $fact_struct->{metadata}{core}{type};

    unless ($guid) {
        Carp::confess "Can't store: no GUID set for fact\n";
    }

    my $json = $self->_json->encode($fact_struct);

    if ( $self->compressed ) {
        $json = compress($json);
    }

    my $s3_object = $self->s3_bucket->object(
        key          => $self->prefix . lc $guid,
#        acl_short    => 'public-read',
        content_type => 'application/json',
    );
    $s3_object->put($json);

    return $guid;
}

# given guid, retrieve it and return it
# type is directory path
# class isa Metabase::Fact::Subclass
sub extract {
    my ( $self, $guid ) = @_;

    my $s3_object = $self->s3_bucket->object( key => $self->prefix . lc $guid );
    return $self->_extract_struct( $s3_object );
}

sub _extract_struct {
  my ( $self, $s3_object ) = @_;

  my $json = $s3_object->get;
  if ( $self->compressed ) {
    $json = uncompress($json);
  }
  my $struct  = $self->_json->decode($json);
  return $struct;
}

# DO NOT lc() GUID
sub delete {
    my ( $self, $guid ) = @_;

    my $s3_object = $self->s3_bucket->object( key => $self->prefix . $guid );
    $s3_object->delete;
}

sub iterator {
  my ($self) = @_;
  return Data::Stream::Bulk::Filter->new(
    stream => $self->s3_bucket->list( { prefix => $self->prefix } ),
    filter => sub {
      return [ map { $self->_extract_struct( $_ ) } @{ $_[0] } ];
    },
  );
}

1;

# ABSTRACT: Metabase storage using Amazon S3
# COPYRIGHT

=for Pod::Coverage::TrustPod store extract delete iterator initialize

=head1 SYNOPSIS

  require Metabase::Archive::S3;
  Metabase::Archive::S3->new(
    access_key_id => 'XXX',
    secret_access_key => 'XXX',
    bucket     => 'acme',
    prefix     => 'metabase/',
    compressed => 0,
  );

=head1 DESCRIPTION

This is an implementation of the L<Metabase::Archive> role using Amazon S3

=head1 USAGE

See L<Metabase::Backend::AWS> for common constructor attributes and see below
for constructor attributes specific to this class.  See L<Metabase::Archive>
and L<Metabase::Librarian> for details on usage.

=attr bucket (required)

S3 bucket name to use for storage

=attr prefix (required)

S3 prefix (within the bucket) to use for storage.  This should be unique
for each Metabase installation.

=attr compressed (deprecated)

Boolean flag indicating whether facts should be compressed prior to S3
storage.  Once facts are stored compressed or not compressed,
this should not be changed for any given Metabase archive.

The default is now true and this attribute deprecated.  It remains to allow
access to older, uncompressed archives.

=cut


lib/Metabase/Backend/AWS.pm
use 5.006;
use strict;
use warnings;

package Metabase::Backend::AWS;
# VERSION

use Moose::Role;
use namespace::autoclean;

has 'access_key_id' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has 'secret_access_key' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

1;

# ABSTRACT: Metabase backend implemented using Amazon Web Services
# COPYRIGHT

=for Pod::Coverage method_names_here

=head1 SYNOPSIS

XXX consolidate synopses from modules

=head1 DESCRIPTION

This distribution provides a backend for L<Metabase> using Amazon Web Services.
There are two modules included, L<Metabase::Index::SimpleDB> and
L<Metabase::Archive::S3>.  They can be used separately or together (see
L<Metabase::Librarian> for details).

The L<Metabase::Backend::AWX> module is a L<Moose::Role> that provides
common attributes and private helpers and is not intended to be used directly.

Common attributes are described further below.

=attr access_key_id

An AWS Access Key ID

=attr secret_access_key

An AWS Secret Access Key matching the Access Key ID

=cut



lib/Metabase/Index/SimpleDB.pm
use 5.006;
use strict;
use warnings;

package Metabase::Index::SimpleDB;
# VERSION

use Moose;
use namespace::autoclean;

use Data::Dumper;
use Data::Stream::Bulk::Callback;
use List::AllUtils qw/sum/;
use SimpleDB::Client;
use Try::Tiny;

with 'Metabase::Backend::AWS';
with 'Metabase::Index' => { -version => 1.000 };

has 'domain' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has 'simpledb' => (
    is      => 'ro',
    isa     => 'SimpleDB::Client',
    lazy    => 1,
    default => sub {
        my $self = shift;
        my $sdb = SimpleDB::Client->new(
            access_key => $self->access_key_id,
            secret_key => $self->secret_access_key
        );
        $sdb->send_request('CreateDomain', { DomainName => $self->domain });
        return $sdb;
    },
);

has consistent => (
  is => 'ro',
  isa => 'Bool',
  default => sub { 0 },
);

sub initialize {}

# XXX not currently used, but available for future
sub _format_int {
    my ($self, $value) = @_;
    $value ||= 0; # init
    return sprintf("%015d",$value+1000000000);
}

sub _get_search_sql {
  my ( $self, $select, $spec ) = @_;

  my ($where, $limit) = $self->get_native_query($spec);

  my $domain = $self->domain;
  my $sql = qq{$select from `$domain` $where};

  return ($sql, $limit);
}

sub add {
    my ( $self, $fact ) = @_;

    Carp::confess("can't index a Fact without a GUID") unless $fact->guid;

    my $metadata = $self->clone_metadata( $fact );

    my $i = 0;
    my @attributes;
    foreach my $key ( sort keys %$metadata ) {
        my $value = $metadata->{$key};
        push @attributes,
            "Attribute.$i.Name"    => $key,
            "Attribute.$i.Value"   => $value;
            # XXX not using replace is an optimization -- dagolden, 2010-04-29
#            "Attribute.$i.Replace" => 'true'; # XXX optimization -- dagolden, 2010-04-29
        $i++;
    }

    my $response = $self->simpledb->send_request(
        'PutAttributes',
        {   DomainName => $self->domain,
            ItemName   => lc $fact->guid,
            @attributes,
        }
    );
}

my $_count_extractor = sub {
  my $response = shift;
  my $items = $response->{SelectResult}{Item};

  # the following may not be necessary as of SimpleDB::Class 1.0000
  $items = [ $items ] unless ref $items eq 'ARRAY';

  my $count = 0;
  for my $i (@$items) {
    next unless $i->{Name} eq 'Domain';
    $count += $i->{Attribute}{Value};
  }

  return [$count], $count;
};

sub count {
  my ( $self, %spec) = @_;
  my $query = "select count(*)";
  my ($sql, $limit) = $self->_get_search_sql($query, \%spec );
  my $cb = Data::Stream::Bulk::Callback->new(
    callback => $self->_generate_fetch_callback($sql, $limit, $_count_extractor)
  );
  return sum $cb->all; 
}

my $_item_extractor = sub {
  my $response = shift;
  my $items = $response->{SelectResult}{Item};

  # the following may not be necessary as of SimpleDB::Class 1.0000
  $items = [ $items ] unless ref $items eq 'ARRAY';

  my $result = [ map { $_->{Name} } @$items ];
  return $result, scalar @$result;
};

sub query {
  my ( $self, %spec) = @_;
  my $query = "select ItemName()";
  my ($sql, $limit) = $self->_get_search_sql($query, \%spec );
  return Data::Stream::Bulk::Callback->new(
    callback => $self->_generate_fetch_callback($sql, $limit, $_item_extractor)
  );
}

sub _generate_fetch_callback {
  my ($self, $sql, $limit, $extractor) = @_;
    # prepare request
  my $request = {
    SelectExpression => $sql,
    ConsistentRead => ($self->consistent ? "true" : "false"),
  };
  my $total_count = 0;
  my $finished = 0;

  return sub {
    return if $finished;
    FETCH: {
      my ($response, $result, $query_count);
      try {
        $response = $self->simpledb->send_request( 'Select', $request );
      } catch {
        die("Got error '$_' from '$sql'");
      };

      if ( exists $response->{SelectResult}{Item} ) {
        ($result, $query_count) = $extractor->($response);
        $total_count += $query_count;
      }

      # If SimpleDB promises more data, then update the request
      if ( exists $response->{SelectResult}{NextToken} ) {
        $request->{NextToken} = $response->{SelectResult}{NextToken};
        # if promised more, but have nothing now, repeat query right away
        redo FETCH unless @$result;
        # if we have more than we need, flag that we're done
        $finished++ if $limit && $total_count >= $limit;
      }
      # No promise of more, so we're done
      else {
        $finished++;
      }

      # return whatever we got
      return $result;
    }
  };
}

# DO NOT lc() GUID
sub delete {
    my ( $self, $guid ) = @_;

    Carp::confess("can't delete without a GUID") unless $guid;

    my $response = $self->simpledb->send_request(
        'DeleteAttributes',
        {   DomainName => $self->domain,
            ItemName   => $guid,
        }
    );
}

#--------------------------------------------------------------------------#
# required by Metabase::Query
#
# ops return closures that define the necessary logic when called
# with hash of index fields
#--------------------------------------------------------------------------#

sub _quote_field {
  my ($self, $field) = @_;
  return qq{`$field`};
}

sub _quote_val {
  my ($self, $value) = @_;
  $value =~ s{"}{""}g;
  return qq{"$value"};
}

sub translate_query {
  my ( $self, $spec ) = @_;

  my (@parts, $limit);

  # where
  if ( defined $spec->{-where} ) {
    push @parts, "where " . $self->dispatch_query_op( $spec->{-where} );
  }

  # order
  if ( defined $spec->{-order} ) {
    my @order_pairs = @{$spec->{-order}};

    Carp::confess("Only one of '-asc' or '-desc' allowed")
      if @{$spec->{-order}} > 2;

    my ($dir, $field) = @{$spec->{-order}};

    Carp::confess("'$field' must be used in WHERE predicates")
      unless $parts[0] && $parts[0] =~ /\Q$field\E/;

    $field = $self->_quote_field( $field );
    $dir =~ s/^-//;
    push @parts, qq{order by $field $dir};
  }

  # limit
  if ( $limit = $spec->{-limit} or $spec->{-order} ) {
    Carp::confess("-limit requires -asc or -desc")
      unless $spec->{-order};
    $limit ||= 2500; # if ordered and no limit, use high default
    push @parts, qq{limit $limit};
  }

  return join( q{ }, @parts ), $limit;
}

sub op_eq {
  my ($self, $field, $val) = @_;
  return $self->_quote_field($field) . " = " . $self->_quote_val($val);
}

sub op_ne {
  my ($self, $field, $val) = @_;
  return $self->_quote_field($field) . " != " . $self->_quote_val($val);
}

sub op_gt {
  my ($self, $field, $val) = @_;
  return $self->_quote_field($field) . " > " . $self->_quote_val($val);
}

sub op_lt {
  my ($self, $field, $val) = @_;
  return $self->_quote_field($field) . " < " . $self->_quote_val($val);
}

sub op_ge {
  my ($self, $field, $val) = @_;
  return $self->_quote_field($field) . " >= " . $self->_quote_val($val);
}

sub op_le {
  my ($self, $field, $val) = @_;
  return $self->_quote_field($field) . " <= " . $self->_quote_val($val);
}

sub op_between {
  my ($self, $field, $low, $high) = @_;
  return $self->_quote_field($field) . " between "
    . $self->_quote_val($low) . " and " . $self->_quote_val($high);
}

sub op_like {
  my ($self, $field, $val) = @_;
  # XXX really should quote/check $val
  return $self->_quote_field($field) . " like " . $self->_quote_val($val);
}

sub op_not {
  my ($self, $pred) = @_;
  my $clause = $self->dispatch_query_op($pred);
  return "NOT ($clause)";
}

sub op_or {
  my ($self, @args) = @_;
  my @predicates = map { $self->dispatch_query_op($_) } @args;
  return join(" or ", map { "($_)" } @predicates);
}

sub op_and {
  my ($self, @args) = @_;
  my @predicates = map { $self->dispatch_query_op($_) } @args;
  return join(" and ", map { "($_)" } @predicates);
}

1;

# ABSTRACT: Metabase index on Amazon SimpleDB
# COPYRIGHT

=for Pod::Coverage::TrustPod add query delete count
initialize translate_query op_eq op_ne op_gt op_lt op_ge op_le op_between op_like
op_not op_or op_and


=head1 SYNOPSIS

  require Metabase::Index::SimpleDB;
  Metabase::Index:SimpleDB->new(
    access_key_id => 'XXX',
    secret_access_key => 'XXX',
    domain     => 'metabase',
  );

=head1 DESCRIPTION

This is an implementation of the L<Metabase::Index> and L<Metabase::Query>
roles using Amazon SimpleDB.

=head1 USAGE

See L<Metabase::Backend::AWS> for common constructor attributes and see below
for constructor attributes specific to this class.  See L<Metabase::Index>,
L<Metabase::Query> and L<Metabase::Librarian> for details on usage.

=attr domain (required)

The SimpleDB domain to store index data in.  This should be unique for
each Metabase installation.

=attr consistent

Whether consistent reads should be used. Default is 0.  Probably
most useful for testing.

=head1 LIMITATIONS

Search queries have limitations based on the underlying SimpleDB search
API.  Specifically:

=over

=item lexicographic comparison

All comparisons are done lexicographically, even when the field or comparison
value appears to be a number.

=item -order

SimpleDB only supports sorting on a single attribute.  Any field used
for sorting must also have been included as part of a query constraint
in the C<-where> expression.

=item -limit

Limit is only supported when a query is sorted.

=back

=cut
